### DISCUSSION TITLE
Does a RaftError always indicate that a node is broken and must shutdown?

### DISCUSSION THREAD

**Discussion by semirix** [databendlabs/openraft#320] [Q&A] [answered]
When a `RaftError` is returned from something like `client_write(request)` I assume `RaftError::ShuttingDown` means that the node is in the process of shutting down and that the error that caused it already happened. If I get any of the other variants in `RaftError` like `RaftError::SnapshotMismatch` should I assume that the entire node is in a broken state and must be shutdown?

**Comment by drmingdrmer** [comment, marked as answer]
I assume you are using `v0.6.4`, in this version `RaftError::SnapshotMismatch` is only used internally thus an application should have no chance of receiving such an error.

I'd recommend using the latest `main` branch in which the errors are better defined:
`client_write()` will return an `ClientWriteError`:
```rust
pub enum ClientWriteError<NID: NodeId> {
    #[error(transparent)]
    ForwardToLeader(#[from] ForwardToLeader<NID>),

    /// When writing a change-membership entry.
    #[error(transparent)]
    ChangeMembershipError(#[from] ChangeMembershipError<NID>),

    #[error(transparent)]
    Fatal(#[from] Fatal<NID>),
}
```

`ClientWriteError` is much more obvious for an application to deal with.

**Reply by semirix** [reply]
Is there a particular commit I should pin my dependencies to until the next version release?

**Reply by drmingdrmer** [reply]
No.

Pinning to the latest `main` would be OK.
The latest main branch has several changes from `0.6.4`, the `example-raft-kv` dir would help.

In the next several weeks there will be some internal refactoring commits that won't affect the API.
After that, there will be several API-changing commits, in order to provide more efficient storage access.

Please be aware of commits start with `Change:`, which brings in breaking changes.
